masonry/src/widget/board.rs

```diff
@@ -0,0 +1,633 @@
+                    );
+                    if ctx.child_needs_layout(shape) {
+                        ctx.run_layout(shape, &BoxConstraints::tight(params.size));
+                    } else {
+                        ctx.skip_layout(shape);
+                    };
+                    ctx.place_child(shape, self.origin);
+                }
+            }
+        }
+
+        bc.max()
+    }
+
+    fn paint(&mut self, _ctx: &mut PaintCtx, _scene: &mut Scene) {}
+
+    fn accessibility_role(&self) -> Role {
+        Role::GenericContainer
+    }
+
+    fn accessibility(&mut self, _ctx: &mut AccessCtx) {}
+
+    fn children_ids(&self) -> SmallVec<[WidgetId; 16]> {
+        self.children
+            .iter()
+            .filter_map(|child| child.widget())
+            .map(|widget_pod| widget_pod.id())
+            .collect()
+    }
+
+    fn make_trace_span(&self) -> Span {
+        trace_span!("Board")
+    }
+}
+
+impl Widget for Shape {
+    fn on_pointer_event(&mut self, _ctx: &mut EventCtx, _event: &PointerEvent) {}
+    fn on_text_event(&mut self, _ctx: &mut EventCtx, _event: &TextEvent) {}
+    fn on_access_event(&mut self, _ctx: &mut EventCtx, _event: &AccessEvent) {}
+    fn on_status_change(&mut self, _ctx: &mut LifeCycleCtx, _event: &StatusChange) {}
+    fn lifecycle(&mut self, _ctx: &mut LifeCycleCtx, _event: &LifeCycle) {}
+
+    fn layout(&mut self, _ctx: &mut LayoutCtx, bc: &BoxConstraints) -> Size {
+        let size = self.shape.bounding_box().size();
+        if !bc.contains(size) {
+            warn!("The shape is oversized");
+        }
+        size
+    }
+
+    fn paint(&mut self, _ctx: &mut PaintCtx, scene: &mut Scene) {
+        scene.fill(
+            self.fill_style,
+            self.transform,
+            &self.fill_brush,
+            self.fill_brush_transform,
+            &self.shape,
+        );
+        scene.stroke(
+            &self.stroke_style,
+            self.transform,
+            &self.stroke_brush,
+            self.stroke_brush_transform,
+            &self.shape,
+        );
+    }
+
+    fn accessibility_role(&self) -> Role {
+        Role::GraphicsSymbol
+    }
+
+    fn accessibility(&mut self, _ctx: &mut AccessCtx) {}
+
+    fn children_ids(&self) -> SmallVec<[WidgetId; 16]> {
+        SmallVec::new()
+    }
+
+    fn make_trace_span(&self) -> Span {
+        trace_span!("Shape")
+    }
+}
+
+// --- MARK: OTHER IMPLS---
+impl BoardParams {
+    /// Create a `BoardParams` with a specific `origin` and `size`.
+    pub fn new(origin: impl Into<Point>, size: impl Into<Size>) -> Self {
+        BoardParams {
+            origin: origin.into(),
+            size: size.into(),
+        }
+    }
+}
+
+impl From<Rect> for BoardParams {
+    fn from(rect: Rect) -> Self {
+        BoardParams {
+            origin: rect.origin(),
+            size: rect.size(),
+        }
+    }
+}
+
+enum Child {
+    Widget(WidgetPod<Box<dyn Widget>>, BoardParams),
+    Shape(WidgetPod<Box<dyn Widget>>),
+}
```

Comment:
Given that `Shape` implements `Widget`, that enum seems redundant.

Reply:
It acts as a type-tag for downcasting in https://github.com/RagibHasin/xilem/blob/991f3e5a90e2ad6d927772d942b6939e5b075c47/masonry/src/widget/board.rs#L316 .